python/pyarrow/_substrait.pyx

```diff
@@ -28,7 +28,7 @@ from pyarrow.includes.libarrow_substrait cimport *
 
 cdef CDeclaration _create_named_table_provider(
     dict named_args, const std_vector[c_string]& names, const CSchema& schema
-):
+) noexcept:
```

Comment:
Are we sure this function cannot raise a Python exception? Its code is definitely non-trivial, so the "solution" here looks more like a bandaid.

In the interest of moving this forward, can you open a bug for this and we'll revisit later?

Reply:
This function executes user-defined python code, which absolutely can raise an exception. In fact, the test cases specifically _do_ raise an exception. The problem is that the C++ functionality explicitly chooses to ignore it and return its own error. The pyarrow binding is designed with this in mind, so adding `noexcept` is mimicking the expected behavior of previous cython versions. I'll file an issue to request refactoring of this feature.

Reply:
Filed https://github.com/apache/arrow/issues/37235

Reply:
I'll see if I can fix this in a separate PR first.

Reply:
> I'll see if I can fix this in a separate PR first.

It would certainly be nice to fix that, but FWIW I don't think that should hold up this PR even longer

Reply:
FYI I haven't attempted #37235. I hope its okay to merge as-is given how long this PR has taken.

Reply:
Can you just add a comment pointing to GH-37235?

Reply:
Will do!